### DISCUSSION TITLE
[WARNING STA-0139] ... 63.3687 is not a float.

### DISCUSSION THREAD

**Discussion by oharboe** [The-OpenROAD-Project/OpenROAD#3314] [General]
I get a lot of these warnings when I run `make gui_final`

This is with the latest ORFS.

What are they?

```
$ make DESIGN_CONFIG=designs/asap7/mock-array-big/Element/config.mk place
```

```
[WARNING STA-0139] /home/oyvind/OpenROAD-flow-scripts/flow/platforms/asap7/lib/asap7sc7p5t_AO_RVT_TT_nldm_211120.lib.gz line 397, 63.3687 is not a float.
[WARNING STA-0139] /home/oyvind/OpenROAD-flow-scripts/flow/platforms/asap7/lib/asap7sc7p5t_AO_RVT_TT_nldm_211120.lib.gz line 397, 79.0805 is not a float.
[WARNING STA-0139] /home/oyvind/OpenROAD-flow-scripts/flow/platforms/asap7/lib/asap7sc7p5t_AO_RVT_TT_nldm_211120.lib.gz line 397, 106.935 is not a float.
[WARNING STA-0139] /home/oyvind/OpenROAD-flow-scripts/flow/platforms/asap7/lib/asap7sc7p5t_AO_RVT_TT_nldm_211120.lib.gz line 397, 156.622 is not a float.
[WARNING STA-0139] /home/oyvind/OpenROAD-flow-scripts/flow/platforms/asap7/lib/asap7sc7p5t_AO_RVT_TT_nldm_211120.lib.gz line 397, 248.596 is not a float.
[WARNING STA-0139] /home/oyvind/OpenROAD-flow-scripts/flow/platforms/asap7/lib/asap7sc7p5t_AO_RVT_TT_nldm_211120.lib.gz line 397, 432.982 is not a float.
[WARNING STA-0139] /home/oyvind/OpenROAD-flow-scripts/flow/platforms/asap7/lib/asap7sc7p5t_AO_RVT_TT_nldm_211120.lib.gz line 406, 0.36 is not a float.
[WARNING STA-0139] /home/oyvind/OpenROAD-flow-scripts/flow/platforms/asap7/lib/asap7sc7p5t_AO_RVT_TT_nldm_211120.lib.gz line 406, 0.72 is not a float.
[WARNING STA-0139] /home/oyvind/OpenROAD-flow-scripts/flow/platforms/asap7/lib/asap7sc7p5t_AO_RVT_TT_nldm_211120.lib.gz line 406, 1.44 is not a float.
[WARNING STA-0139] /home/oyvind/OpenROAD-flow-scripts/flow/platforms/asap7/lib/asap7sc7p5t_AO_RVT_TT_nldm_211120.lib.gz line 406, 2.88 is not a float.
[WARNING STA-0139] /home/oyvind/OpenROAD-flow-scripts/flow/platforms/asap7/lib/asap7sc7p5t_AO_RVT_TT_nldm_211120.lib.gz line 406, 5.76 is not a float.
[WARNING STA-0139] /home/oyvind/OpenROAD-flow-scripts/flow/platforms/asap7/lib/asap7sc7p5t_AO_RVT_TT_nldm_211120.lib.gz line 406, 11.52 is not a float.
[WARNING STA-0139] /home/oyvind/OpenROAD-flow-scripts/flow/platforms/asap7/lib/asap7sc7p5t_AO_RVT_TT_nldm_211120.lib.gz line 406, 23.04 is not a float.
[WARNING STA-0139] /home/oyvind/OpenROAD-flow-scripts/flow/platforms/asap7/lib/asap7sc7p5t_AO_RVT_TT_nldm_211120.lib.gz line 415, 12.8576 is not a float.
[WARNING STA-0139] /home/oyvind/OpenROAD-flow-scripts/flow/platforms/asap7/lib/asap7sc7p5t_AO_RVT_TT_nldm_211120.lib.gz line 415, 15.7364 is not a float.
[WARNING STA-0139] /home/oyvind/OpenROAD-flow-scripts/flow/platforms/asap7/lib/asap7sc7p5t_AO_RVT_TT_nldm_211120.lib.gz line 415, 21.3969 is not a float.
[WARNING STA-0139] /home/oyvind/OpenROAD-flow-scripts/flow/platforms/asap7/lib/asap7sc7p5t_AO_RVT_TT_nldm_211120.lib.gz line 415, 32.5931 is not a float.
```

**Comment by vijayank88** [comment]
@oharboe 
I ran with latest commit and not able to get similar logs.

**Comment by maliberty** [comment]
The report says 'make place' and 'make gui_final' which are not a compatible pair.   I can't reproduce this either even running all the way through.

**Comment by oharboe** [comment]
```
rm -rf results/
make DESIGN_CONFIG=designs/asap7/mock-array-big/Element/config.mk place
make DESIGN_CONFIG=designs/asap7/mock-array-big/Element/config.mk gui_place
```

```
$ make DESIGN_CONFIG=designs/asap7/mock-array-big/Element/config.mk gui_place
[INFO][FLOW] Using platform directory ./platforms/asap7
[INFO-FLOW] ASU ASAP7 - version 2
Default PVT selection: BC
ODB_FILE=./results/asap7/mock-array-big_Element/base/3_place.odb /home/oyvind/OpenROAD-flow-scripts/tools/install/OpenROAD/bin/openroad -gui  ./scripts/gui.tcl
Warning: Ignoring XDG_SESSION_TYPE=wayland on Gnome. Use QT_QPA_PLATFORM=wayland to run on Wayland anyway.
This plugin does not support propagateSizeHints()
OpenROAD v2.0-8114-gaa1f8060d 
This program is licensed under the BSD-3 license. See the LICENSE file for details.
Components of this program may be licensed under more restrictive licenses which must be honored.
[WARNING STA-0139] ./platforms/asap7/lib/asap7sc7p5t_AO_RVT_FF_nldm_211120.lib.gz line 82, 0.72 is not a float.
[WARNING STA-0139] ./platforms/asap7/lib/asap7sc7p5t_AO_RVT_FF_nldm_211120.lib.gz line 82, 1.44 is not a float.
[WARNING STA-0139] ./platforms/asap7/lib/asap7sc7p5t_AO_RVT_FF_nldm_211120.lib.gz line 82, 2.88 is not a float.
[WARNING STA-0139] ./platforms/asap7/lib/asap7sc7p5t_AO_RVT_FF_nldm_211120.lib.gz line 82, 5.76 is not a float.
[WARNING STA-0139] ./platforms/asap7/lib/asap7sc7p5t_AO_RVT_FF_nldm_211120.lib.gz line 82, 11.52 is not a float.
[WARNING STA-0139] ./platforms/asap7/lib/asap7sc7p5t_AO_RVT_FF_nldm_211120.lib.gz line 82, 23.04 is not a float.
[WARNING STA-0139] ./platforms/asap7/lib/asap7sc7p5t_AO_RVT_FF_nldm_211120.lib.gz line 82, 46.08 is not a float.
[WARNING STA-0139] ./platforms/asap7/lib/asap7sc7p5t_AO_RVT_FF_nldm_211120.lib.gz line 96, 0.72 is not a float.
[WARNING STA-0139] ./platforms/asap7/lib/asap7sc7p5t_AO_RVT_FF_nldm_211120.lib.gz line 96, 1.44 is not a float.
[WARNING STA-0139] ./platforms/asap7/lib/asap7sc7p5t_AO_RVT_FF_nldm_211120.lib.gz line 96, 2.88 is not a float.
[WARNING STA-0139] ./platforms/asap7/lib/asap7sc7p5t_AO_RVT_FF_nldm_211120.lib.gz line 96, 5.76 is not a float.
```

**Reply by vijayank88** [reply]
@oharboe 
You're trying to forcefully remove all `results` directory. It lost all previous stage results.
How you're able to run `place` stage directly?

**Reply by oharboe** [reply]
make place will work from a full clean and create all its dependencies, but make gui_final requires everything to built.

**Comment by oharboe** [comment]
@maliberty Aha! This is a locale issue. In their wisdom a hundred years ago, Norwegians decided to use "," to separate integers from fractions in floating point... So "12.34" is not a valid float in Norwegian locale, rather "12,34" is.

The evidence that this is the problem is that setting `LC_NUMERIC="en_GB.UTF-8"` fixes the problem.

In this case, the locale is decided by the format that is being parsed, it is not a human readable/writeable number.

Let's see if this works: https://github.com/Pinata-Consulting/OpenSTA/tree/lock-down-strtof-locale

Yep :-)

This also explains why we don't see this when we run from docker, only local install.

**Comment by maliberty** [comment]
Brazil has the same issue.  https://github.com/The-OpenROAD-Project/OpenROAD/blob/6920ac530b6959f742e75b7f55062b1642730b14/src/Main.cc#L217  attempts to solve this.  Do none of the locales exist on your systems?

**Reply by oharboe** [reply]
```
   for (auto locale : locales) {
+    std::cerr << "Trying " << locale << std::endl;
     if (std::setlocale(LC_ALL, locale) != nullptr) {
+      std::cerr << "Found! " << locale << std::endl;
       break;
     }
   }
```

=>

Trying en_US.UTF-8
Found! en_US.UTF-8

**Reply by maliberty** [reply]
If it is set to en_US.UTF-8 then it should be using '.' as the decimal separator not ','.  Do you not observe that?

**Reply by maliberty** [reply]
Are you setting some of the more narrow LC_ vars?

**Reply by oharboe** [reply]
Not explicitly, this is what I get out of the box on a Norwegian Ubuntu machine:

```
$ export | grep LC
declare -x LC_ADDRESS="nb_NO.UTF-8"
declare -x LC_IDENTIFICATION="nb_NO.UTF-8"
declare -x LC_MEASUREMENT="nb_NO.UTF-8"
declare -x LC_MONETARY="nb_NO.UTF-8"
declare -x LC_NAME="nb_NO.UTF-8"
declare -x LC_NUMERIC="nb_NO.UTF-8"
declare -x LC_PAPER="nb_NO.UTF-8"
declare -x LC_TELEPHONE="nb_NO.UTF-8"
declare -x LC_TIME="nb_NO.UTF-8"
```

Also, one more experiment:

```
   for (auto locale : locales) {
+    std::cerr << "Trying " << locale << std::endl;
     if (std::setlocale(LC_ALL, locale) != nullptr) {
+      std::cerr << "Found! " << locale << std::endl;
       break;
     }
   }
+  const char *string ="2.234";
+  char *end;
+  float value = strtof(string, &end);
+  if (*end && !isspace(*end)) {
+    std::cerr << "failed to parse: " << end << std::endl;
+  }
+  std::cerr << "parsed " << value << std::endl;
```

Trying en_US.UTF-8
Found! en_US.UTF-8
parsed 2.234

Rabbit hole...

**Reply by oharboe** [reply]
![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/2798822/f3a3d6b0-7d07-4f1c-b9c9-81f3dbcaa998)

![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/2798822/c589f9e8-82cc-4bf7-b9ce-4c99452542ca)

**Reply by oharboe** [reply]
![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/2798822/06ba1654-4065-47a4-a9d4-a3e12f7d9cf4)

**Reply by oharboe** [reply]
![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/2798822/ec7f23b4-b672-4ef8-b645-5ec4e9d26fc9)

**Reply by oharboe** [reply]
Hmmm... setlocale() is probably called to read the current locale.

By the time it is in liberty parsing, the locale is still what it was set to.

![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/2798822/d90b5493-6df3-4d15-a2e8-e84982363608)

**Reply by oharboe** [reply]
So the locale is set, but the result from strtof() still parses "Norwegian" numbers...

![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/2798822/75ebb1d6-5783-40a5-83a1-ccd634fcf4e0)

![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/2798822/989faa2c-3cf8-42b6-93be-565c94c35564)


Some more experiments:

```
Console is in 'commands' mode, prefix expressions with '?'.
Attached to process 301940
expr (char*) setlocale(0, NULL)
(char *) $0 = 0x000055f26e47b610 "en_US.UTF-8"
expr (float) strtof("234.1234", NULL)
(float) $1 =

 234
expr (float) strtof("234,1234", NULL)

(float) $2 = 234.123398
call setlocale(6, "C")
(char *) $3 =

 0x00007f0acdfb75ff "C"
expr (float) strtof("234.1234", NULL)
(float) $4 = 234.123398

call setlocale(6, "C")
(char *) $5 = 0x00007f0acdfb75ff "C"

expr (float) strtof("234.1234", NULL)
(float)

$6 = 234.123398
call setlocale(6, "en_US.UTF-8")
(char *) $7 =

 0x000055f26e4e9730 "en_US.UTF-8"
expr (float) strtof("234.1234", NULL)
(float) $8 =

 234.123398
```


So.... that's weird...

**Reply by oharboe** [reply]
![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/2798822/bd41dbc7-fc0f-414b-8d9d-1f817d7c5473)

**Reply by oharboe** [reply]
![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/2798822/bad866ac-34d7-4368-bc35-a7f53dc9a288)

**Reply by oharboe** [reply]
![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/2798822/eb0ab21e-8d91-49e7-9569-4f06c887fa89)

**Comment by maliberty** [comment]
Would you try your strtof check along with getlocale after these various points to try to see what's going on.  I think this works in Brazil so I'm trying to understand what's different here.  @luis201420 any insights?

**Reply by luis201420** [reply]
I have this error when I execute with GUI activated, to solve it I add these lines to set `LC_ALL` in the first part of `tclAppInit` function in **Main.cc**
```
std::array locales = {"en_US.UTF-8", "C.UTF-8", "C"};
  for (auto locale : locales) {
    if (std::setlocale(LC_ALL, locale) != nullptr) {
      break;
    }
}
```
I found that in the main function `LC_ALL` is set but after calling the `Tcl_Main` function again it is reset

**Reply by oharboe** [reply]
I see the same thing... setlocal() is called perhaps once per major framework? Qt, Tcl, Python, etc.... Everybody wants to help out...

**Comment by maliberty** [comment]
Perhaps it relates to LC_NUMERIC or another.  Can you try unsetting all LC_* but LC_ALL in your environment to see if that matters?

**Reply by oharboe** [reply]
LC_ALL is not set.

**Reply by maliberty** [reply]
In the c++ code try setting all the locale variables rather than just LC_ALL and see if that helps.

**Comment by oharboe** [comment]
@maliberty @luis201420 With this change, I no longer get the warnings.

This is just a quick hack, I think it would be better if someone who knows what is going on and are aware of other edgecases articulates the pull request :-)

```
diff --git a/src/gui/src/gui.cpp b/src/gui/src/gui.cpp
index 355aa6842..7d79d99a3 100644
--- a/src/gui/src/gui.cpp
+++ b/src/gui/src/gui.cpp
@@ -1141,6 +1141,16 @@ int startGui(int& argc,
   }
   main_window->show();
 
+  // This avoids problems with locale setting dependent
+  // C functions like strtod (e.g. 0.5 vs 0,5).
+  std::array locales = {"en_US.UTF-8", "C.UTF-8", "C"};
+  for (auto locale : locales) {
+    if (std::setlocale(LC_ALL, locale) != nullptr) {
+      break;
+    }
+  }
+
+
   gui->setLogger(open_road->getLogger());
 
   main_window->setDatabase(open_road->getDb());
```

**Comment by maliberty** [comment]
@luis201420 will you look at this since you can reproduce it

**Comment by oharboe** [comment]
I guess this has been identified as a bug in OpenROAD, given that there are frameworks that set the locale after OpenROAD has set the locale.
